### PR TITLE
Fix for building on mac with Gnu compilers; allow third party package…

### PR DESCRIPTION
**Pull request by BarrySmith**
…s to be provided in alternative locations

Mac does not support or need either  -Wl,-all_load or -Wl,--whole-archive

Remove nonsense about MUMPS Fortran MPI library, that is handled through the normal mpicxx, etc

Add support for
-DMA57_DIR
-DCOIN_DIR
-DAMPL_DIR
-DMETIS_DIR
-DMUMPS_DIR
 so external libraries can be built seperately and not provided all in the same directory

make test and the individual examples for NLP in Test seem to run correctly

### CODE REVIEW

**Comment by cnpetra** [comment]
thanks @BarrySmith much appreciated, building on mac was always problematic

it looks good me as long as @michel2323 is ok with removing the "nonsense" MUMPS Fortran lib. I think it was put in there with a reason ...

**Comment by BarrySmith** [comment]
Actually I am now working to revert and put back in the nonsense MPI library :-). In my use case it is not needed but since you are using the mpicxx compiler to link the shared libraries (and it doesn't usually know about the Fortran mpi libraries that only mpifc provides) the libraries do need to be provided explicitly. Once it is back the CI travis tests should work correctly.

 Sorry for the noise (I wish PR systems didn't contact reviewers until after all the tests passed but I am screaming against the wind since they all do which is totally bizarre to me, I've been fighting with Jed as we construct our new CI system for PETSc so it won't bother reviewers until the tests are passed but it is difficult 1) to convince Jed that it is reasonable to not contact the reviewers until the tests pass and 2) to actually get CI/PR systems to work this way.

**Comment by michel2323** [comment]
@kibaekkim Would you be willing to give it a try on the mac?

**Comment by BarrySmith** [comment]
It would be good if Kibaek confirms it works for him.

 Note that I don't use the sh scripts for building the third party packages (I have PETSc build them; and use the PIPS cmake to build PIPS NLP only) so I don't know if the sh scripts work on the Mac. I could not get PIPS non-NLP to build due to boost issues and don't think it is worth pursing that.

**Comment by michel2323** [comment]
The `.sh` scripts are not portable at all. I just added them as a 'guide'.

**Comment by kibaekkim** [comment]
I used the `.sh` scripts (`build_3rdparty.sh` and `build_pips.sh`). Here are the results:
- By default, Mac uses `clang`, and it is not compatible to `openmp`. `CC` and `CXX` have to be set manually to use gnu compiler. If you run `build_3rdparty.sh`, it sets `CC` but not `CXX`. So anyway I had to set `CXX`.
- Running `build_pips.sh` compiles everything, but it fails at compiling `PIPSS` and `PIPS_IPM`. It complains about undefined symbols related to MPI. So I turned on `PIPS_NLP` only. It cannot find `mpi.h`. Did I miss something for MPI?

**Comment by BarrySmith** [comment]
Are you using CXX = mpicxx? Or directly the compiler? Most Cmake usages are built around assuming using the compiler wrappers. If you are using mpicxx can you list or attach the exact error?

**Comment by kibaekkim** [comment]
@BarrySmith The error I got must have been related to compiling with "multi-threads". To reproduce and report the error, I ran `make` again, and it went through....

**Comment by kibaekkim** [comment]
I tried compiling `PIPS_IPM` and `PIPS_S`. Both failed, although I don't think that is related to Barry's PR. I did not dig into that. Here are the error screen I saw for each.

## PIPS_IPM error
```
ld: warning: option -noall_load is obsolete and being ignored
Undefined symbols for architecture x86_64:
  "PardisoSchurSolver::PardisoSchurSolver(SparseSymMatrix*)", referenced from:
      sLinsysLeaf::sLinsysLeaf<PardisoSchurSolver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchurSolver*) in libooqpstoch.a(sFactoryAugSchurLeaf.C.o)
      sLinsysLeaf::sLinsysLeaf<PardisoSchurSolver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchurSolver*) in libooqpstoch.a(sFactoryAugComm2SchurLeaf.C.o)
  "PardisoSchur32Solver::PardisoSchur32Solver(SparseSymMatrix*)", referenced from:
      sLinsysLeaf::sLinsysLeaf<PardisoSchur32Solver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchur32Solver*) in libooqpstoch.a(sFactoryAugSchurLeaf.C.o)
  "typeinfo for PardisoSchurSolver", referenced from:
      sLinsysLeafSchurSlv::addTermToDenseSchurCompl(sData*, DenseSymMatrix&)       in libooqpstoch.a(sLinsysLeafSchurSlv.C.o)
  "_ma27ad_", referenced from:
      Ma27SolverBase::firstCall()      in libooqpgensparse.a(Ma27Solver.C.o)
  "_ma27bd_", referenced from:
      Ma27SolverBase::matrixChanged()       in libooqpgensparse.a(Ma27Solver.C.o)
  "_ma27cd_", referenced from:
      Ma27SolverBase::basicSolve(double*, int) in libooqpgensparse.a(Ma27Solver.C.o)
      Ma27Solver::solve(OoqpVector&)       in libooqpgensparse.a(Ma27Solver.C.o)
  "_ma27id_", referenced from:
      Ma27Solver::Ma27Solver(SparseSymMatrix*) in libooqpgensparse.a(Ma27Solver.C.o)
      Ma27Solver::Ma27Solver(SparseSymMatrix*) in libooqpgensparse.a(Ma27Solver.C.o)
      Ma27SolverBase::Ma27SolverBase(int, int) in libooqpgensparse.a(Ma27Solver.C.o)
      Ma27SolverBase::Ma27SolverBase(int, int) in libooqpgensparse.a(Ma27Solver.C.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [PIPS-IPM/libpipsipm-shared.dylib] Error 1
make[1]: *** [PIPS-IPM/CMakeFiles/pipsipm-shared.dir/all] Error 2
make: *** [all] Error 2
```

## PIPS_S error
```
Undefined symbols for architecture x86_64:
  "ClpModel::copyNames(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)", referenced from:
      ClpBALPInterface::ClpBALPInterface(stochasticInput&, BAContext&, BALPSolverInterface<ClpBALPInterface>::solveType) in libClpBALPInterface.a(ClpBALPInterface.cpp.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [clpSMPS] Error 1
make[1]: *** [PIPS-S/CMakeFiles/clpSMPS.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by cnpetra** [comment]
> I tried compiling `PIPS_IPM` and `PIPS_S`. Both failed, although I don't think that is related to Barry's PR. I did not dig into that. Here are the error screen I saw for each.
> 
> ## PIPS_IPM error
> ```
> ld: warning: option -noall_load is obsolete and being ignored
> Undefined symbols for architecture x86_64:
>   "PardisoSchurSolver::PardisoSchurSolver(SparseSymMatrix*)", referenced from:
>       sLinsysLeaf::sLinsysLeaf<PardisoSchurSolver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchurSolver*) in libooqpstoch.a(sFactoryAugSchurLeaf.C.o)
>       sLinsysLeaf::sLinsysLeaf<PardisoSchurSolver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchurSolver*) in libooqpstoch.a(sFactoryAugComm2SchurLeaf.C.o)
>   "PardisoSchur32Solver::PardisoSchur32Solver(SparseSymMatrix*)", referenced from:
>       sLinsysLeaf::sLinsysLeaf<PardisoSchur32Solver>(sFactory*, sData*, OoqpVector*, OoqpVector*, OoqpVector*, OoqpVector*, PardisoSchur32Solver*) in libooqpstoch.a(sFactoryAugSchurLeaf.C.o)
>   "typeinfo for PardisoSchurSolver", referenced from:
>       sLinsysLeafSchurSlv::addTermToDenseSchurCompl(sData*, DenseSymMatrix&)       in libooqpstoch.a(sLinsysLeafSchurSlv.C.o)
>   "_ma27ad_", referenced from:
>       Ma27SolverBase::firstCall()      in libooqpgensparse.a(Ma27Solver.C.o)
>   "_ma27bd_", referenced from:
>       Ma27SolverBase::matrixChanged()       in libooqpgensparse.a(Ma27Solver.C.o)
>   "_ma27cd_", referenced from:
>       Ma27SolverBase::basicSolve(double*, int) in libooqpgensparse.a(Ma27Solver.C.o)
>       Ma27Solver::solve(OoqpVector&)       in libooqpgensparse.a(Ma27Solver.C.o)
>   "_ma27id_", referenced from:
>       Ma27Solver::Ma27Solver(SparseSymMatrix*) in libooqpgensparse.a(Ma27Solver.C.o)
>       Ma27Solver::Ma27Solver(SparseSymMatrix*) in libooqpgensparse.a(Ma27Solver.C.o)
>       Ma27SolverBase::Ma27SolverBase(int, int) in libooqpgensparse.a(Ma27Solver.C.o)
>       Ma27SolverBase::Ma27SolverBase(int, int) in libooqpgensparse.a(Ma27Solver.C.o)
> ld: symbol(s) not found for architecture x86_64
> collect2: error: ld returned 1 exit status
> make[2]: *** [PIPS-IPM/libpipsipm-shared.dylib] Error 1
> make[1]: *** [PIPS-IPM/CMakeFiles/pipsipm-shared.dir/all] Error 2
> make: *** [all] Error 2
> ```
> 
> ## PIPS_S error
> ```
> Undefined symbols for architecture x86_64:
>   "ClpModel::copyNames(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)", referenced from:
>       ClpBALPInterface::ClpBALPInterface(stochasticInput&, BAContext&, BALPSolverInterface<ClpBALPInterface>::solveType) in libClpBALPInterface.a(ClpBALPInterface.cpp.o)
> ld: symbol(s) not found for architecture x86_64
> collect2: error: ld returned 1 exit status
> make[2]: *** [clpSMPS] Error 1
> make[1]: *** [PIPS-S/CMakeFiles/clpSMPS.dir/all] Error 2
> make: *** [all] Error 2
> ```

most likely messed up logic/conditionals

**Comment by michel2323** [comment]
> most likely messed up logic/conditionals in cmake scripts

Are you trying on a Mac? There PIPS-IPM and PIPS-S still won't work. Or is your error on Linux?

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,12 +89,11 @@ if (BUILD_PIPS_IPM OR BUILD_PIPS_NLP)
...
 endif()
 
-# include different "whole archive" linking options depending on compiler
-if (CMAKE_CXX_COMPILER_ID MATCHES "^(Apple)?Clang$")
-  set(WHOLE_ARCHIVE "-Wl,-all_load")
-  set(NO_WHOLE_ARCHIVE "-Wl,-noall_load")
-  message(STATUS "SETTING HAVE_CLANG")
-  set(HAVE_CLANG TRUE)
+# include different "whole archive" linking options depending on system
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
```
Comment (kibaekkim) [inline CMakeLists.txt]: How is different from `if(APPLE)`? Inside this `if` condition, there is `if(APPLE)` too.

File: CMakeLists.txt [hunk 2]
```diff
@@ -89,12 +89,11 @@ if (BUILD_PIPS_IPM OR BUILD_PIPS_NLP)
...
+  message(STATUS "Detected Darwin turning off --whole-archive")
+  set(WHOLE_ARCHIVE "")
```
Comment (kibaekkim) [inline CMakeLists.txt]: Although I could successfully compile `PIPS-NLP` on my mac, the previous setting `set(WHOLE_ARCHIVE "-Wl,-all_load")` also worked for me.

**Comment by kibaekkim** [comment]
> > most likely messed up logic/conditionals in cmake scripts
> 
> Are you trying on a Mac? There PIPS-IPM and PIPS-S still won't work. Or is your error on Linux?

The errors were from my Mac.

File: CMakeLists.txt (hunk 2 above)
Comment (BarrySmith) [reply CMakeLists.txt]: Perhaps you have an older version of Xcode (that does the linking; Gnu only does the compiling) and it still supports this option? My linker said it didn't exist. (It's kind of a weird option anyways and probably in most circumstances doesn't matter either way).

**Comment by BarrySmith** [comment]
Sorry, I may not have been clear. I gave up trying to get the build through for anything but NLP on the Mac; I couldn't get it working. I guess the main question is if I broke anything on Linux where presumably all of them should build fine.

**Comment by cnpetra** [comment]
Ok, I’ll give a try on Linux before merging just to be make sure

**Pull request merged**